alchemist/alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/geometry/euclidean/twod/AbstractConvexPolygon.kt

```diff
@@ -0,0 +1,190 @@
+ * as described in the file LICENSE in the Alchemist distribution's top directory.
+ */
+
+package it.unibo.alchemist.model.implementations.geometry.euclidean.twod
+
+import it.unibo.alchemist.model.implementations.geometry.SegmentsIntersectionType
+import it.unibo.alchemist.model.implementations.geometry.intersection
+import it.unibo.alchemist.model.implementations.geometry.vertices
+import it.unibo.alchemist.model.implementations.positions.Euclidean2DPosition
+import it.unibo.alchemist.model.interfaces.geometry.Vector2D
+import it.unibo.alchemist.model.interfaces.geometry.euclidean.twod.ConvexPolygon
+import it.unibo.alchemist.model.interfaces.geometry.euclidean.twod.Segment2D
+import java.awt.Shape
+import java.lang.IllegalArgumentException
+import java.lang.IllegalStateException
+import kotlin.math.min
+
+/**
+ * An abstract [ConvexPolygon] providing a convexity test.
+ */
+abstract class AbstractConvexPolygon : ConvexPolygon {
+
+    override fun liesOnBoundary(vector: Euclidean2DPosition): Boolean = edges().any { it.contains(vector) }
+
+    override fun containsBoundaryIncluded(vector: Euclidean2DPosition): Boolean =
+        contains(vector) || liesOnBoundary(vector)
+
+    override fun containsBoundaryExcluded(vector: Euclidean2DPosition): Boolean =
+        contains(vector) && !liesOnBoundary(vector)
+
+    override fun contains(shape: Shape): Boolean = shape.vertices().all { containsBoundaryIncluded(it) }
+
+    /*
+     * It's important that intersects(Shape) does not consider adjacent shapes as intersecting.
+     */
+    override fun isAdjacentTo(other: ConvexPolygon): Boolean = !intersects(other.asAwtShape()) &&
+        (other.vertices().any { liesOnBoundary(it) } || vertices().any { other.liesOnBoundary(it) })
+
+    override fun isAdjacentTo(shape: Shape): Boolean = isAdjacentTo(fromShape(shape).orElseThrow {
+        IllegalArgumentException("given shape is not convex polygonal")
+    })
+
+    override fun intersects(segment: Segment2D<Euclidean2DPosition>): Boolean {
+        if (containsBoundaryExcluded(segment.first) || containsBoundaryExcluded(segment.second)) {
+            return true
+        }
+        val intersections = edges().map { intersection(it, segment) }
+        return when {
+            intersections.any { it.type == SegmentsIntersectionType.SEGMENT } -> false
+            else -> intersections.mapNotNull { it.point.orElse(null) }.distinct().size > 1
+        }
+    }
+
+    override fun closestEdgeTo(segment: Segment2D<Euclidean2DPosition>): Segment2D<Euclidean2DPosition> =
+        edges().minWith(
+            compareBy({
+                it.distanceTo(segment)
+            }, {
+                val segmentDistanceToEdgeBounds = segment.distanceTo(it.first) + segment.distanceTo(it.second)
+                val edgeDistanceToSegmentBounds = it.distanceTo(segment.first) + it.distanceTo(segment.second)
+                min(segmentDistanceToEdgeBounds, edgeDistanceToSegmentBounds)
```

Comment:
This lambda could become a private extension function of `Segment2D`, possibly in the companion

Reply:
I refactored it in two private extension functions in `AbstractConvexPolygon`'s companion.